core/lib/zksync_core/src/sync_layer/gossip/storage.rs

```diff
@@ -0,0 +1,351 @@
+use zksync_dal::{ConnectionPool, StorageProcessor};
+use zksync_types::{Address, MiniblockNumber};
+
+use super::{buffered::ContiguousBlockStore, conversions::sync_block_to_consensus_block};
+use crate::sync_layer::{
+    fetcher::{FetchedBlock, FetcherCursor},
+    sync_action::{ActionQueueSender, SyncAction},
+};
+
+#[derive(Debug)]
+struct CursorWithCachedBlock {
+    inner: FetcherCursor,
+    maybe_last_block_in_batch: Option<FetchedBlock>,
+}
+
+impl From<FetcherCursor> for CursorWithCachedBlock {
+    fn from(inner: FetcherCursor) -> Self {
+        Self {
+            inner,
+            maybe_last_block_in_batch: None,
+        }
+    }
+}
+
+impl CursorWithCachedBlock {
+    fn advance(&mut self, block: FetchedBlock) -> Vec<Vec<SyncAction>> {
+        let mut actions = Vec::with_capacity(2);
+        if let Some(mut prev_block) = self.maybe_last_block_in_batch.take() {
+            prev_block.last_in_batch = prev_block.l1_batch_number != block.l1_batch_number;
+            actions.push(self.inner.advance(prev_block));
+        }
+
+        // We take advantage of the fact that the last block in a batch is a *fictive* block that
+        // does not contain transactions. Thus, any block with transactions cannot be last in an L1 batch.
+        let can_be_last_in_batch = block.transactions.is_empty();
+        if can_be_last_in_batch {
+            self.maybe_last_block_in_batch = Some(block);
+            // We cannot convert the block into actions yet, since we don't know whether it seals an L1 batch.
+        } else {
+            actions.push(self.inner.advance(block));
+        }
+        actions
+    }
+}
+
+#[derive(Debug)]
+pub(super) struct PostgresBlockStorage {
+    pool: ConnectionPool,
+    actions: ActionQueueSender,
+    block_sender: watch::Sender<BlockNumber>,
+    cursor: Mutex<CursorWithCachedBlock>,
+}
+
+impl PostgresBlockStorage {
+    pub fn new(pool: ConnectionPool, actions: ActionQueueSender, cursor: FetcherCursor) -> Self {
+        let current_block_number = cursor.next_miniblock.0.saturating_sub(1).into();
+        Self {
+            pool,
+            actions,
+            block_sender: watch::channel(BlockNumber(current_block_number)).0,
+            cursor: Mutex::new(cursor.into()),
+        }
+    }
+
+    pub async fn run_background_tasks(&self, ctx: &ctx::Ctx) -> StorageResult<()> {
+        const POLL_INTERVAL: time::Duration = time::Duration::milliseconds(50);
+        loop {
+            let sealed_miniblock_number = self
+                .sealed_miniblock_number()
+                .await
+                .map_err(StorageError::Database)?;
+            self.block_sender.send_if_modified(|number| {
+                if *number != sealed_miniblock_number {
+                    *number = sealed_miniblock_number;
+                    true
+                } else {
+                    false
+                }
+            });
+            if ctx.sleep(POLL_INTERVAL).await.is_err() {
```

Comment:
nit: to indicate that the expecte error is cancellation, you can write it as: `if let Err(ctx::Canceled) = ctx.sleep() { ... }`